ospfd/ospf_asbr.c

```diff
@@ -329,3 +331,885 @@ void ospf_redistribute_withdraw(struct ospf *ospf, uint8_t type,
+					     "Ospf external route hash");
+	return aggr;
+}
+
+static void ospf_aggr_handle_external_info(void *data)
+{
+	struct external_info *ei = (struct external_info *)data;
+	struct ospf_external_aggr_rt *aggr = NULL;
+	struct ospf *ospf = NULL;
+	struct ospf_lsa *lsa = NULL;
+
+	ei->aggr_route = NULL;
+
+	ei->to_be_processed = true;
+
+	if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+		zlog_debug("%s: Handle extrenal route(%pI4/%d)", __func__,
+			   &ei->p.prefix, ei->p.prefixlen);
+
+	ospf = ospf_lookup_instance(ei->instance);
+
+	assert(ospf);
+
+	if (!ospf_redistribute_check(ospf, ei, NULL))
+		return;
+
+	aggr = ospf_external_aggr_match(ospf, &ei->p);
+	if (aggr) {
+		lsa = ospf_originate_summary_lsa(ospf, aggr, ei);
+		return;
+	}
+
+	lsa = ospf_external_info_find_lsa(ospf, &ei->p);
+	if (lsa)
+		ospf_external_lsa_refresh(ospf, lsa, ei, LSA_REFRESH_FORCE, 1);
+	else
+		lsa = ospf_external_lsa_originate(ospf, ei);
+}
+
+static void ospf_aggr_unlink_external_info(void *data)
+{
+	struct external_info *ei = (struct external_info *)data;
+
+	ei->aggr_route = NULL;
+
+	ei->to_be_processed = true;
+}
+
+void ospf_external_aggregator_free(struct ospf_external_aggr_rt *aggr)
+{
+	if (OSPF_EXTERNAL_RT_COUNT(aggr))
+		hash_clean(aggr->match_extnl_hash,
+			   (void *)ospf_aggr_unlink_external_info);
+
+	if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+		zlog_debug("%s: Release the aggregator Address(%pI4/%d)",
+			   __func__, &aggr->p.prefix, aggr->p.prefixlen);
+	hash_free(aggr->match_extnl_hash);
+	aggr->match_extnl_hash = NULL;
+
+	XFREE(MTYPE_OSPF_EXTERNAL_RT_AGGR, aggr);
+}
+
+static void ospf_external_aggr_add(struct ospf *ospf,
+				   struct ospf_external_aggr_rt *aggr)
+{
+	struct route_node *rn;
+
+	if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+		zlog_debug("%s: Adding Aggregate route to Aggr table (%pI4/%d)",
+			   __func__, &aggr->p.prefix, aggr->p.prefixlen);
+	rn = route_node_get(ospf->rt_aggr_tbl, (struct prefix *)&aggr->p);
+	if (rn->info)
+		route_unlock_node(rn);
+	else
+		rn->info = aggr;
+}
+
+static void ospf_external_aggr_delete(struct ospf *ospf, struct route_node *rn)
+{
+	struct ospf_external_aggr_rt *aggr = rn->info;
+
+	if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+		zlog_debug("%s: Deleting Aggregate route (%pI4/%d)", __func__,
+			   &aggr->p.prefix, aggr->p.prefixlen);
+
+	/* Sent a Max age LSA if it is already originated. */
+	if (CHECK_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED)) {
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug("%s: Flushing Aggregate route (%pI4/%d)",
+				   __func__, &aggr->p.prefix,
+				   aggr->p.prefixlen);
+		ospf_external_lsa_flush(ospf, 0, &aggr->p, 0);
+	}
+
+	rn->info = NULL;
+	route_unlock_node(rn);
+	route_unlock_node(rn);
+}
+
+struct ospf_external_aggr_rt *
+ospf_extrenal_aggregator_lookup(struct ospf *ospf, struct prefix_ipv4 *p)
+{
+	struct route_node *rn;
+
+	rn = route_node_lookup(ospf->rt_aggr_tbl, (struct prefix *)p);
+	if (rn) {
+		route_unlock_node(rn);
+		return rn->info;
```

Comment:
This is a possible use after free.  Please grab the rn->info before the unlock

Reply:
Yes , Addressed.